hphp/runtime/vm/jit/vasm-lower.cpp

```diff
@@ -33,6 +33,55 @@ namespace HPHP { namespace jit {
 
 ///////////////////////////////////////////////////////////////////////////////
 
+namespace detail {
+
+jit::vector<int> computeDominatingConstraints(Vunit& unit) {
+  jit::vector<int> vregConstraints(unit.blocks.size(), -1);
+
+  // Build reverse postorder list of blocks
+  auto const rpo = sortBlocks(unit);
+
+  // Calculate each block's dominating vreg constraint. The lowerers will use
+  // this as the initial vreg constraint for each block, adjusting as they
+  // encounter vregallow/vregrestrict during lowering.
+  vregConstraints[rpo[0]] = 0;
+  for (auto b : rpo) {
+    // Iterate through the instructions updating the constraint accordingly.
+    auto con = vregConstraints[b];
+    for (size_t i = 0; i < unit.blocks[b].code.size(); ++i) {
+      auto& inst = unit.blocks[b].code[i];
+      if (inst.op == Vinstr::vregrestrict) {
+        con++;
+      } else if (inst.op == Vinstr::vregallow) {
+        con--;
+      }
+    }
+    assert_flog(con >= 0,
+		"Dominating constraint can't be unknown.");
```

Comment:
After follow-up discussions: Will be changing uninitialized to INT_MAX and changing the constraint level to restrict for < 0 and allow for >=0. Will be changing boolean to also be an int level. Also changing the names of the Vinstrs to better indicate that they just adjust the vreg constraint level.